### ISSUE TITLE
OSX Yosemite breaks Spoof-Mac

### ISSUE THREAD

**Issue by Parsav** [feross/SpoofMAC#36]
You can run the program, it sucessfuly turns off and on the wifi, but doesn't change the address

**Comment by applocalizations** [comment]
Has a fix evolved from the discussion for setting the MAC address?

**Comment by asmeurer** [comment]
The Wikipedia article no longer seems to have a reference for this (or I didn't find it when skimming through it).

**Comment by prisconapoli** [comment]
Take a look at the image (on the rigth side) of this link http://en.wikipedia.org/wiki/MAC_address#Address_details.

In few words, the less two significant bits in the first octet of a MAC Address are special flags, used to define some aspects of the address.

The less significant bit is used to differentiate among Unicast (0) or Multicast(1) address, while the second one is used to differentiate among Universally(0) or Locally(1) administrated address.

Generally, in a personal computer, a NIC is set for accepting only unicast (0) address. There are not restriction about universal or local address.

You can use the second bit to communicate that you are no longer using the manufacturer’s  burned-in-address (BIA) and that MAC address has been changed locally. So be careful about this bit too, if you want to make a mac spoofing in a proper way.

Hope this help,
P.

**Comment by applocalizations** [comment]
These are already well known principles about MAC addresses. The question
is implementing these principles in the program so that the `set` command
works.

2015-03-27 2:31 GMT-07:00 Prisco Napoli <notifications@github.com>:

**Comment by pteek** [comment]
How about a verify function that will check and correct to always return a valid mac address to spoof?

It can be called just before setting a new mac weather it is random or user given.

**Comment by applocalizations** [comment]
whether* An `if else` should work but it might be easier to let the code
run until a valid one results, lest the code become large (i.e., a
corrective function might not be necessary).

2015-03-27 8:12 GMT-07:00 pteek <notifications@github.com>:

**Comment by pteek** [comment]
For arkanoid87 above, he was spoofing a manually specified address which had the multicast bit set, so we have to handle the case when a user makes this mistake. 

We can either change the MAC address silently or warn the user. In any case, we will need a validate function (5 lines).

**Comment by TkTech** [comment]
The set function is working as intended - a verifier that *warns* but doesn't forbid the user is probably best. If I want to explicitly set a mac address nothing should prevent me from doing so.

For the randomize function, it should definitely generate one that verifies.

**Comment by pteek** [comment]
Okay, I'll write one verifying function. Make randomize always return a valid mac address. And add warning code to the script.

One thing, should the randomly generated address be LA or UA?

**Comment by pteek** [comment]
Randomize has a fixed OUI prefix in the vmware range. It always generates a unicast address.

We only need to warn the user if he enters a multicast address.

**Comment by applocalizations** [comment]
I agree. The function manipulate MAC addresses when a user enters a
multicast address is unnecessary so warning the user makes the most sense.

2015-03-27 9:01 GMT-07:00 pteek <notifications@github.com>:

**Comment by pteek** [comment]
PR merged.

**Comment by applocalizations** [comment]
Awesome . . .

2015-03-27 9:37 GMT-07:00 pteek <notifications@github.com>:

**Comment by prisconapoli** [comment]
well done :)

**Comment by feross** [comment]
Just published 2.0.4 with this improvement.

**Comment by jpalcantara** [comment]
Hello,
  I installed the following pip install SpoofMAC and  easy_install SpoofMAC and I also wanted to automate this.
if I run sudo spoof-mac.py randomize en1 it does what it does
then I run spoof-mac.py list and shows what it's supposed to.
But I installed the Startup Installation .  And restarted my system and there is no changes on the WIFI card and also is the local.macspoof.plist supposed to be empty?
